Change ragged column lengths to offsets

jeromekelleher opened the pull request:
Given that an actual release is around the corner, we need to fix up any remaining warts on the tables API so that we're not stuck with them. A big wart on the current API IMO is that we are dealing with ragged columns by passing the length of each item as a separate array. This PR changes this to passing an array of offsets instead. This is a more efficient approach, and will allow for easier access of tables by row, as well as simplifying the tree sequence C code.

To be specific, if we have a ragged column like ``site.ancestral_state``, we currently have two arrays, ``ancestral_state`` and ``ancestral_state_length``. I propose change this to ``ancestral_state`` and ``ancestral_state_offset``. Concretely, if we had 3 ancestral states ['A', 'CCAC', 'G'], the current encoding is
```python
ancestral_state = [A, C, C, A, C, G]
ancestral_state_length = [1, 4, 1]
```
and the proposed encoding is
```python
ancestral_state = [A, C, C, A, C, G]
ancestral_state_offset = [0, 1, 5, 6]
```
Note that 

1. ancestral_state_offset[0] is always equal to zero
2. The length of ancestral_state_offset is num_rows + 1

This formulation means that we can find the ancestral state for a given row ``j`` using
```python
a = ancestral_state[ancestral_state_offset[j]: ancestral_state_offset[j + 1]]
```

The code here is a WIP on the C side of things, where there are quite few tedious changes. However, I do think that this is the right approach for ragged columns, and we should fix this before we release 0.5.0. 

Any thoughts @petrelharp?

jeromekelleher commented:
> My thoughts: ￼ ￼ ￼
> Too bad about the tedium.

OK, cool, I'll chip away at this so. A small part of me was hoping you'd come up with a good reason for why this is a terrible idea and I wouldn't have to go through the tedium!

petrelharp commented:
Sadly, no.  At least with the switch to edges we won't want to do this for
`children` also...

On Wed, Nov 29, 2017 at 9:15 AM, Jerome Kelleher <notifications@github.com>
wrote:

> My thoughts: ￼ ￼ ￼
> Too bad about the tedium.
>
> OK, cool, I'll chip away at this so. A small part of me was hoping you'd
> come up with a good reason for why this is a terrible idea and I wouldn't
> have to go through the tedium!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jeromekelleher/msprime/pull/322#issuecomment-347930856>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA_26eFzHfOssG9C1_lOsozytUV634zbks5s7ZFJgaJpZM4QvM9a>
> .
>

jeromekelleher commented:
OK, I think this is covers the changes needed to replace `x_length` columns with `x_offset` columns in the tables API. This also affects the low-level tree sequence code, and makes dealing with these columns substantially less horrible.

The practical upshot of this change for users is that it will break people's code (those that are tracking master anyway and using sites/mutation tables), and will make their tree sequence files unreadable (unless we write a converter). In the interest of minimising this disruption, I think I should add the remaining changes that we need before merging this PR. These are:

1. Change the (unused) EdgeTable.name column to ``metadata``.
2. Add a `metadata` column to the site table.
3. Add a ProvenanceTable.

These changes were discussed in #284.

I need 1, 2 and 3 for the stuff I'm doing with real data. I'm happy to add a ``metadata`` column to MutationTable too, but it seems less urgent to me and easy to tack on in future if it's needed.

Any thoughts @petrelharp, @ashander, @molpopgen?

petrelharp commented:
Sounds good.  Just so I'm clear, the point of the ProvenanceTable is to say
what version of msprime produced the tree sequence?

Do you have a README accumulating of important changes?  It'd be easy to at
least describe what you'd need to do to convert.

On Tue, Dec 5, 2017 at 9:34 AM, Jerome Kelleher <notifications@github.com>
wrote:

> OK, I think this is covers the changes needed to replace x_length columns
> with x_offset columns in the tables API. This also affects the low-level
> tree sequence code, and makes dealing with these columns substantially less
> horrible.
>
> The practical upshot of this change for users is that it will break
> people's code (those that are tracking master anyway and using
> sites/mutation tables), and will make their tree sequence files unreadable
> (unless we write a converter). In the interest of minimising this
> disruption, I think I should add the remaining changes that we need before
> merging this PR. These are:
>
>    1. Change the (unused) EdgeTable.name column to metadata.
>    2. Add a metadata column to the site table.
>    3. Add a ProvenanceTable.
>
> These changes were discussed in #284
> <https://github.com/jeromekelleher/msprime/issues/284>.
>
> I need 1, 2 and 3 for the stuff I'm doing with real data. I'm happy to add
> a metadata column to MutationTable too, but it seems less urgent to me
> and easy to tack on in future if it's needed.
>
> Any thoughts @petrelharp <https://github.com/petrelharp>, @ashander
> <https://github.com/ashander>, @molpopgen <https://github.com/molpopgen>?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jeromekelleher/msprime/pull/322#issuecomment-349380075>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA_26bmGOkrzJ7hQ25TbS_YO7Scx46tCks5s9X69gaJpZM4QvM9a>
> .
>

jeromekelleher commented:
> Sounds good.  Just so I'm clear, the point of the ProvenanceTable is to say what version of msprime produced the tree sequence?

It's to record the version and environment information for *every* program that has produced a particular tree sequence. Suppose I ran a large simulation with ftprime, and then made a tree sequence of a subset of it. Then, the provenance table should have two rows (or maybe three): rows. One recording version information about ftprime, and parameters used to generate the simulation, and a second recording the version of msprime used to run ``simplify``. It's to give an audit trail for where a tree sequence came from, and to make them reproducible (in principle at least; I wouldn't get too hung up on actually making it happen too soon).

I'm feeling the need for this at the moment, as I generate tree sequences from data, (with lots of intermediate states), and it would be very handy to be able to see where they came from.

> Do you have a README accumulating of important changes?  It'd be easy to at least describe what you'd need to do to convert.

No --- it'd be quicker to write a converter to be honest. Are there many people working with version 9 files (that can't be easily reproduced) do you think? I've been careful to make sure that released file format versions are upgradeable, but I've got through 6 or 7 versions since the last release.

ashander commented:
> It's to record the version and environment information for every program that has produced a particular tree sequence. Suppose I ran a large simulation with ftprime, and then made a tree sequence of a subset of it. Then, the provenance table should have two rows (or maybe three): rows. One recording version information about ftprime, and parameters used to generate the simulation, and a second recording the version of msprime used to run simplify. It's to give an audit trail for where a tree sequence came from, and to make them reproducible (in principle at least; I wouldn't get too hung up on actually making it happen too soon).

This seems useful. I'm currently running into reproducibility issues, which I think are downstream from simplify. But while chasing them down it has crossed my mind that simplify could cause trouble. Knowing were a given sequence comes from would be nice to eliminate some potential issues, so +1! The format in #284 seems reasonable.

> Are there many people working with version 9 files (that can't be easily reproduced) do you think? I've been careful to make sure that released file format versions are upgradeable, but I've got through 6 or 7 versions since the last release.

Not me. I can reproduce or use old msprime versions where needed. 

Speaking of this, though, it would be nice to tag or otherwise document the file format versions even on the master branch. (At least going forward.) This would aid in finding the last commit that works with a given file format. I had to do this recently with git bisect and `python -c 'import msprime; msprime.load("mytree.hdf5")' but it was tedious. Now that I look I see going backwards through `git blame` on [these lines](https://github.com/ashander/msprime/blame/8d659b6c407a04a63b7d2e261a8a911a8346afb2/lib/msprime.h#L40-L41) would've been a better bet

petrelharp commented:
> Are there many people working with version 9 files 

Nothing that can't be easily re-done that I know of...

jeromekelleher commented:
> Speaking of this, though, it would be nice to tag or otherwise document the file format versions even on the master branch. (At least going forward.) This would aid in finding the last commit that works with a given file format. I had to do this recently with git bisect and python -c 'import msprime; msprime.load("mytree.hdf5")' but it was tedious. Now that I look I see going backwards through git blame on these lines would've been a better bet

Yes, that is annoying, sorry about that. These tags are a good idea, and could be done retrospectively too using git blame on the lines you pointed out.

jeromekelleher commented:
Phew, this has turned into a biggie. I think I'm nearly done now though, just need to add the `metadata` column to the SiteTable and that should be it. 

Provenance handling is hugely improved now, and should actually be useable (although the details of what's recorded needs to be worked out).

jeromekelleher commented:
OK, I think this is finally ready to go.  This updates the HDF5 file format version to 10,  and so any existing version 9 files will have to be recreated. Pinging @petrelharp, @ashander, @molpopgen and @hyanwong to alert you to this.

Here are the changes:

1. Changed all string columns to use the colname, colname_offsets convention as described above. 
2. Added ProvenanceTable, which contains two string columns, ``timestamp`` and ``record``.
3. Changed ``Node.name`` to ``Node.metadata``, and added ``Site.metadata`` and ``Mutation.metadata``.
4. Changed ``sort_tables`` to return the ``node_map`` array rather than pass it in as an argument (if you don't use node_map, then you're code isn't affected). This was motivated by the difficulties with releasing the GIL.

There is various work left to be done on the metadata and provenance side of things (mostly: update the text versions and document the new APIs), but the basic infrastructure is now in place and I'd like to merge.

petrelharp commented:
yay!

On Tue, Dec 12, 2017 at 7:03 AM, Jerome Kelleher <notifications@github.com>
wrote:

> OK, I think this is finally ready to go. This updates the HDF5 file format
> version to 10, and so any existing version 9 files will have to be
> recreated. Pinging @petrelharp <https://github.com/petrelharp>, @ashander
> <https://github.com/ashander>, @molpopgen <https://github.com/molpopgen>
> and @hyanwong <https://github.com/hyanwong> to alert you to this.
>
> Here are the changes:
>
>    1. Changed all string columns to use the colname, colname_offsets
>    convention as described above.
>    2. Added ProvenanceTable, which contains two string columns, timestamp
>    and record.
>    3. Changed Node.name to Node.metadata, and added Site.metadata and
>    Mutation.metadata.
>    4. Changed sort_tables to return the node_map array rather than pass
>    it in as an argument (if you don't use node_map, then you're code isn't
>    affected). This was motivated by the difficulties with releasing the GIL.
>
> There is various work left to be done on the metadata and provenance side
> of things (mostly: update the text versions and document the new APIs), but
> the basic infrastructure is now in place and I'd like to merge.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jeromekelleher/msprime/pull/322#issuecomment-351077288>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA_26XSOHdamMjpudF3Uye4iiZiAlssgks5s_pWogaJpZM4QvM9a>
> .
>

jeromekelleher commented:
OK, merging. One more warning that version 9 files will no longer work @petrelharp, @ashander, @molpopgen and @hyanwong!
